flume-ng-sinks/flume-http-sink/src/main/java/org/apache/flume/sink/http/HttpSink.java

```diff
@@ -0,0 +1,408 @@
+
+    connectTimeout = context.getInteger("connectTimeout",
+        DEFAULT_CONNECT_TIMEOUT);
+
+    if (connectTimeout <= 0) {
+      throw new IllegalArgumentException(
+          "Connect timeout must be a non-zero and positive");
+    }
+    LOG.info("Using connect timeout : " + connectTimeout);
+
+    requestTimeout = context.getInteger("requestTimeout",
+        DEFAULT_REQUEST_TIMEOUT);
+
+    if (requestTimeout <= 0) {
+      throw new IllegalArgumentException(
+          "Request timeout must be a non-zero and positive");
+    }
+    LOG.info("Using request timeout : " + requestTimeout);
+
+    acceptHeader = context.getString("acceptHeader", DEFAULT_ACCEPT_HEADER);
+    LOG.info("Using Accept header value : " + acceptHeader);
+
+    contentTypeHeader = context.getString("contentTypeHeader",
+        DEFAULT_CONTENT_TYPE);
+    LOG.info("Using Content-Type header value : " + contentTypeHeader);
+
+    defaultBackoff = context.getBoolean("defaultBackoff", true);
+    LOG.info("Channel backoff by default is " + defaultBackoff);
+
+    defaultRollback = context.getBoolean("defaultRollback", true);
+    LOG.info("Transaction rollback by default is " + defaultRollback);
+
+    defaultIncrementMetrics = context.getBoolean("defaultIncrementMetrics",
+        false);
+    LOG.info("Incrementing metrics by default is " + defaultIncrementMetrics);
+
+    parseConfigOverrides("backoff", context, backoffOverrides);
+    parseConfigOverrides("rollback", context, rollbackOverrides);
+    parseConfigOverrides("incrementMetrics", context,
+        incrementMetricsOverrides);
+
+    if (this.sinkCounter == null) {
+      this.sinkCounter = new SinkCounter(this.getName());
+    }
+
+    connectionBuilder = new ConnectionBuilder();
+  }
+
+  @Override
+  public final void start() {
+    LOG.info("Starting HttpSink");
+    sinkCounter.start();
+  }
+
+  @Override
+  public final void stop() {
+    LOG.info("Stopping HttpSink");
+    sinkCounter.stop();
+  }
+
+  @Override
+  public final Status process() throws EventDeliveryException {
+    Status status = null;
+    OutputStream outputStream = null;
+
+    Channel ch = getChannel();
+    Transaction txn = ch.getTransaction();
+    txn.begin();
+
+    try {
+      Event event = ch.take();
+
+      byte[] eventBody = null;
+      if (event != null) {
+        eventBody = event.getBody();
+      }
+
+      if (eventBody != null && eventBody.length > 0) {
+        sinkCounter.incrementEventDrainAttemptCount();
+        LOG.debug("Sending request : " + new String(event.getBody()));
+
+        try {
+          HttpURLConnection connection = connectionBuilder.getConnection();
+
+          outputStream = connection.getOutputStream();
+          outputStream.write(eventBody);
+          outputStream.flush();
+          outputStream.close();
+
+          int httpStatusCode = connection.getResponseCode();
+          LOG.debug("Got status code : " + httpStatusCode);
+
+          connection.getInputStream().close();
```

Comment:
Just my 0.02, rather than create then close HttpURLConnection every time, could we try to re-use HttpURLConnection if the previous one is still good (or even keep it alive on purpose)?